protocols/kad/src/jobs.rs

```diff
@@ -0,0 +1,406 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Periodic (background) jobs.
+//!
+//! ## Record Persistence & Expiry
+//!
+//! To ensure persistence of records in the DHT, a Kademlia node
+//! must periodically (re-)publish and (re-)replicate its records:
+//!
+//!   1. (Re-)publishing: The original publisher or provider of a record
+//!      must regularly re-publish in order to prolong the expiration.
+//!
+//!   2. (Re-)replication: Every node storing a replica of a record must
+//!      regularly re-replicate it to the closest nodes to the key in
+//!      order to ensure the record is present at these nodes.
+//!
+//! Re-publishing primarily ensures persistence of the record beyond its
+//! initial TTL, for as long as the publisher stores (or provides) the record,
+//! whilst (re-)replication primarily ensures persistence for the duration
+//! of the TTL in the light of topology changes. Consequently, replication
+//! intervals should be shorter than publication intervals and
+//! publication intervals should be shorter than the TTL.
```

Comment:
That is just a matter of naming. In order to distinguish the hourly re-publishing done by every node from the 24h re-publishing done by the node that is the "original publisher", I call the former re-replication. The replication of values to a newly joined node can of course be given separate attention, but since that is also covered by the hourly re-republication, I didn't want to over-complicate things.

Reply:
I guess my choice of naming is partly derived from the choice of naming of the constants for the intervals in [this design specification](http://xlattice.sourceforge.net/components/protocol/kademlia/specs.html#constants).